Improve usability and fix some issues with hotkey remap dialog

dragunoff opened the pull request:
Fix a number of usability issues and make some changes and improvements.

~~I opted not to inline `dialog-hotkey.yaml` into `settings.yaml` because that complicates things with the chrome logic.~~ Ultimately the whole logic and yaml moved into `SettingsLogic` and `settings.yaml`.

~~There are a bunch of commits but I can squash/reorder if needed.~~

Closes #16812

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -486,11 +480,14 @@ Action InitHotkeysPanel(Widget panel)
 					var remapDialogRoot = panel.Get("HOTKEY_DIALOG_ROOT");
-					var remapDialogPlaceholder = panel.GetOrNull("HOTKEY_DIALOG_PLACEHOLDER");
+
 					foreach (var t in types)
```
Please add braces around these other blocks so the inner one isn't left hanging.

pchote reviewed (commented):
In the long term we would ideally rewrite the `HotkeyDialogLogic` to work like `MapPreviewLogic` rather than having to destroy and recreate the panel each time a new key is pressed. Right now, however, the priority needs to be getting something that works well so we can ship a release.

One inline code comment, and a couple of more general requests:
* Can you please move the `HOTKEY_DIALOG` definition from `dialog-hotkey.yaml` into `settings.yaml`? It doesn't need to be inline, keep it as its own block, just have it defined in the same file so its clear that it is a dialog within the settings window.
* The focus behaviour for the hotkey widget is weird. It is focused by default when opening the panel, and is refocused when selecting any other hotkey. Most of the time it remains focused when you change the hotkey value, but sometimes it doesn't. The behaviour of the reset vs cancel buttons is quite unintuitive - they do basically the same thing, but cancel only works while the widget is focused and the change hasn't been committed.

  IMO it would be a lot more intuitive if the hotkey widget remained focused all of the time and changes were always committed immediately. The Cancel button could then be removed, then the entry widget and the remaining two buttons can be move to the right and the hotkey type label moved down so it is on the same line. This height of the bottom panel can then be reduced to make more space for the top scroll panel.

dragunoff commented:
>  The focus behaviour for the hotkey widget is weird. It is focused by default when opening the panel, and is refocused when selecting any other hotkey. Most of the time it remains focused when you change the hotkey value, but sometimes it doesn't. The behaviour of the reset vs cancel buttons is quite unintuitive - they do basically the same thing, but cancel only works while the widget is focused and the change hasn't been committed.
>
>  IMO it would be a lot more intuitive if the hotkey widget remained focused all of the time and changes were always committed immediately. The Cancel button could then be removed, then the entry widget and the remaining two buttons can be move to the right and the hotkey type label moved down so it is on the same line. This height of the bottom panel can then be reduced to make more space for the top scroll panel.
>

The widget remains focused only in case the new value is not valid (has a duplicate). Focus is lost if the new value is valid and that serves as feedback that it was commited. I'm not sure about an always-focused widget but I'll try it.

I agree that the cancel button is a bit unintuitive. It is actually a "remove-focus-if-hotkey-is-valid" button because I wanted the user to have the option to cancel an active rebinding. It works much better if there is an explicit "Save/OK" button which was the case in my first implementation. But in the case of an inline panel I'd be fine with removing the cancel button.

dragunoff commented:
Updated and rebased. 

* Moved the dialog inside `settings.yaml`
* Removed cancel button
* Rearranged the panel
* Cleaned up some redundant logic

I've decided against an always focused widget - I think it's good to have feedback that an entry has been commited. 

I'm not very fond of the overall layout of the hotkeys page but this is probably the best we can get at this point. If the flexbox layout engine gets in then a proper redesign can be considered.

![ts](https://user-images.githubusercontent.com/1355810/62403974-af77d280-b599-11e9-9274-cbd19cdda8bc.png)

![cnc](https://user-images.githubusercontent.com/1355810/62404005-cfa79180-b599-11e9-84c4-ee9a2e740c9d.png)

pchote commented:
Why does TD use icons, but the other mods text?

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -139,45 +139,38 @@ static void BindHotkeyPref(HotkeyDefinition hd, HotkeyManager manager, Widget te
-				foreach (var sibling in siblings)
+			remapButton.OnClick = () =>
+			{
+				foreach (var sibling in parent.Children)
```
Rather than manually walking all the siblings to manually toggle their state, define a `HotkeyDefinition selected` in the class, bind `button.IsHighlighted = () => selected == hd;` once during widget creation, then just update `selected` on select.

dragunoff replied:
This leads to CS0120 because `BindHotkeyPref` is a static method. Should I make the property or the method static to get around this?

pchote replied:
Make `BindHotkeyPref` not static instead!

dragunoff replied:
Done.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -486,11 +480,18 @@ Action InitHotkeysPanel(Widget panel)
-								BindHotkeyPref(hd, modData.Hotkeys, template, hotkeyList, remapDialogRoot, remapDialogPlaceholder);
+							{
+								BindHotkeyPref(hd, modData.Hotkeys, template, hotkeyList, remapDialogRoot, isFirstHotkey);
+								isFirstHotkey = false;
```
`isFirstHotkey` can the also hopefully go away (just set `selected` to the first hotkey here/above).

dragunoff replied:
I find it easier with a `isFirstHotkey` boolean. Otherwise the code has to check if it's the first element in each of the nested loops. I'm open to suggestions of course.

dragunoff commented:
> 
> 
> Why does TD use icons, but the other mods text?

TD is the odd one when it comes to UI. And it has appropriate icons for both buttons. But I'd be fine switching to text only.

matjaeck commented:
The color that indicates conflicting hotkey bindings is not updated when you change one binding (the untouched binding stays colored, although no longer conflicting). Otherwise :+1:

dragunoff commented:
> 
> 
> The color that indicates conflicting hotkey bindings is not updated when you change one binding (the untouched binding stays colored, although no longer conflicting). Otherwise 👍

Good catch. Should I fix this here or in a follow up?

matjaeck commented:
I'd say do it here if it's an easy fix.

dragunoff commented:
Updated with a fix for the color that indicates conflicting hotkey bindings.

abcdefg30 commented:
Imho the remap dialogue shouldn't have auto-focus when opening the hotkey menu. It would be better if it only took focus when you selected a hotkey or entered the field yourself.
> lawANDorder: i tend to agree with abcdefg30 here. when i only want to look up a hotkey it feels a bit dangerous that I'm suddenly in "edit mode"

dragunoff commented:
Yeah, I also think it's a little stressful to jump straight into edit mode. What should be the initial state then? Select first hotkey without focusing  it OR have a blank state without any key selected (like it is before this PR).

abcdefg30 commented:
I think both is fine, but the consensus on IRC was "Select first hotkey without focusing it".

dragunoff commented:
Rebased and updated:

- Reorder and squash some of the commits
- Hightlight the first hotkey but without focusing
- Fix the reset button for the whole panel (was totaly broken)
- Remove the dialog when switching to a different panel (otherwise the widget retains focus)

There is minor problem with the last fix - when switching away and back to the hotkeys panel, the remap dialog would be gone. I could not find a solution to that...

pchote commented:
> There is minor problem with the last fix - when switching away and back to the hotkeys panel, the remap dialog would be gone

This is normally done inside the action returned by `InitHotkeysPanel` (see `InitDisplayPanel` for a proper example), which is why having the logic controlling the dialog split into its own widget and chrome logic makes life really difficult.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -489,16 +484,31 @@ Action InitHotkeysPanel(Widget panel)
+			return () =>
+			{
+				selectedHotkey = null;
+				remapDialogRoot.RemoveChildren();
```
Replace this with something like
```csharp
// Force the hotkey selection widget, if active, to yield focus
// HACK: we don't manage these widgets, so have to make a fragile guess
var w = Ui.KeyboardFocusWidget;
while (w != null)
{
	if (w == remapDialogRoot)
		Ui.KeyboardFocusWidget.YieldKeyboardFocus();

	w = w.Parent;
}
```
to work around the disappearing dialog issue.

dragunoff committed:
Move hotkey dialog logic into SettingsLogic, fix bugs and improve usability of the dialog

dragunoff commented:
Updated as discussed on IRC - moved the whole dialog logic into `SettingsLogic`. Fixed a number of bugs along the way and simplified the code a bit.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -745,5 +748,78 @@ void MakeMouseFocusSettingsLive()
+
+			if (hotkeyEntryWidget.Key == selectedHotkeyDefinition.Default || (!hotkeyEntryWidget.Key.IsValid() && !selectedHotkeyDefinition.Default.IsValid()))
+			{
+				defaultHotkeyNotice.Visible = !duplicateHotkeyNotice.Visible;
```
We generally prefer to make widgets observe other state rather than toggle visibility/disabled/etc manually. This really helps avoid edge cases and weird UI bugs.

In these cases it could be good to define a `enum HotkeyStatus { Default, Custom, Duplicate, Invalid }` (not sure exactly what cases we need) and an `HotkeyStatus activeHotkeyStatus;`. We can then bind e.g. `resetHotkeyButton.IsDisabled = () => activeHotkeyStatus != HotkeyStatus.Default` and so on.

This has a nice side-effect of keeping the button / label widgets local, so they won't need to be defined at the class scope.

dragunoff replied:
I found that a single enum did not cut it because a hotkey can be `Default` (or not) and `Duplicate` at the same time. I used two booleans instead - `isHotkeyValid` and `isHotkeyDefault`.

dragunoff commented:
Updated to address pchote's comments. I added a fixup commit, will squash it once the PR is approved.

pchote reviewed (commented):
This is looking a lot better now, thanks :)

Just a few more optimizations, then this should be good to go..

dragunoff commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -134,50 +140,31 @@ static void BindHotkeyPref(HotkeyDefinition hd, HotkeyManager manager, Widget te
-					if (button != null)
-						button.IsHighlighted = () => false;
-				}
+				return modData.Hotkeys.GetFirstDuplicate(hd.Name, modData.Hotkeys[hd.Name].GetValue(), hd) != null ?
```
But I guess the color won't auto-update once there are no more duplicates?

pchote replied:
Is `isHotkeyValid` not going to auto-update when there are no more duplicates?

dragunoff replied:
It would update only for the current hotkey. But we need to also update the color of the past duplicate.

pchote replied:
I see. In that case, just fix the `ChromeMetrics` bit here and we can deal with the duplicates in a followup PR - we could add a `Duplicate` flag (or maybe a list of other hotkeys?) to the `HotkeyDefinition` class, which can be updated by the `HotkeyManager` when other keys are changed.

dragunoff replied:
As a work-around could we update the color manually only when a hotkey is saved (i.e. loop all the buttons and re-check for duplicates)?

pchote replied:
Adding the extra state to `HotkeyDefinition` seems like a more elegant solution IMO. I don't mind filing that PR if you're not sure quite how to tackle it.

dragunoff commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -745,5 +747,68 @@ void MakeMouseFocusSettingsLive()
+
+			var defaultHotkeyNotice = panel.Get<LabelWidget>("DEFAULT_NOTICE");
+			defaultHotkeyNotice.TextColor = ChromeMetrics.Get<Color>("NoticeInfoColor");
+			defaultHotkeyNotice.IsVisible = () => isHotkeyValid && isHotkeyDefault;
```
I guess this can be removed as it is a bit redundant. There are notices for duplicate and custom hotkey, so not displaying any notice means that this is the default.

dragunoff commented:
Updated to address latest comments. Also shortened some local var names.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -134,50 +140,34 @@ static void BindHotkeyPref(HotkeyDefinition hd, HotkeyManager manager, Widget te
 
-				if (remapDialogPlaceholder != null)
-					remapDialogPlaceholder.Visible = false;
+			var hotkeyValidColor = ChromeMetrics.Get<Color>("HotkeyColorInvalid");
```
Variable names seem to be inverted?

dragunoff replied:
:man_facepalming: Fixed.

pchote commented on OpenRA.Mods.Common/Widgets/Logic/SettingsLogic.cs:
```diff
@@ -745,5 +750,68 @@ void MakeMouseFocusSettingsLive()
+			var duplicateNotice = panel.Get<LabelWidget>("DUPLICATE_NOTICE");
+			duplicateNotice.TextColor = ChromeMetrics.Get<Color>("NoticeErrorColor");
+			duplicateNotice.IsVisible = () => !isHotkeyValid;
+			var duplicateNoticeText = new CachedTransform<HotkeyDefinition, string>(hd => (hd != null) ? duplicateNotice.Text.F(hd.Description) : duplicateNotice.Text);
```
Nit: redundant () around the hd != null check

dragunoff replied:
Fixed.

pchote reviewed (dismissed):
LGTM. Please squash the commits now.

reaperrr reviewed (approved):
LGTM

reaperrr commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/7deabcfafd62fb67615ab4920e7024b0d03be74f...cc9eda62c420a6ab3118c8880205230e0b71c794
